<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>exercise-07a.knit</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Phylogenetics</a>
</li>
<li>
  <a href="schedule.html">Schedule</a>
</li>
<li>
  <a href="projects.html">Projects</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">




</div>


<p><br />
</p>
<div
id="exploring-model-assumptions-and-the-fossilized-birth-death-process"
class="section level1">
<h1>Exploring model assumptions and the fossilized birth-death
process</h1>
<p>This exercise will walk through key assumptions of the fossilized
birth-death (FBD) process using the R package FossilSim.</p>
<p>The FBD process is a joint model of diversification and sampling,
first described in Stadler (2010). It can be used to calculate the
probability of observing a phylogeny with fossils, as extinct leaves or
sampled ancestors.</p>
<p>The model is statistically coherent for the analysis of phylogenetic
and fossil data because it features explicit assumptions about the
underlying data generating processes. This means we can forward
simulate, as well as perform inference, under the model, which is a
major advantage for understanding the model properties and
assumptions.</p>
<p>Install the necessary R packages.</p>
<pre class="r"><code>install.packages(&quot;TreeSim&quot;)
install.packages(&quot;FossilSim&quot;)</code></pre>
<div id="model-parameters" class="section level2">
<h2>Model parameters</h2>
<p>The model has four parameters:</p>
<ul>
<li>lineage birth rate (<span
class="math inline">\(\lambda\)</span>)</li>
<li>lineage death rate (<span class="math inline">\(\mu\)</span>)</li>
<li>fossil recovery rate (<span
class="math inline">\(\psi\)</span>)</li>
<li>probability of extant species sampling (<span
class="math inline">\(\rho\)</span>)</li>
</ul>
<p>The birth and death rate parameters are part of the diversification
component of the model and describe how the tree grows over time, while
the fossil recovery and extant species sampling parameters describe how
the tree is <em>incompletely</em> sampled over time.</p>
<p>The diversification and sampling processes can be considered
separately, which is convenient for both simulation and understanding
different the two components (diversification and sampling) of the
model.</p>
</div>
<div id="the-diversification-process" class="section level2">
<h2>The diversification process</h2>
<p>First we’ll simulate trees, and explore the role of the birth and
death parameters.</p>
<p>In the following we will simulate birth-death trees conditioned on
the number of tips <span class="math inline">\(n\)</span>, using the
<code>TreeSim</code> package function <code>sim.taxa.bd</code>. The
function simulates birth and death events forward in time until there
are <span class="math inline">\(n\)</span> co-existing lineages.</p>
<p>Let’s give it a go.</p>
<pre class="r"><code># set the random number generator seed to generate the same results using the same code
set.seed(1234)

# simulate a tree using TreeSim conditioned on tip number
lambda = 0.1
mu = 0.05
tips = 10
t = TreeSim::sim.bd.taxa(n = tips, numbsim = 1, lambda = lambda, mu = mu)[[1]]
# t is an object of class phylo
t</code></pre>
<pre><code>## 
## Phylogenetic tree with 10 tips and 9 internal nodes.
## 
## Tip labels:
##   t5, t6, t1, t9, t3, t8, ...
## 
## Rooted; includes branch lengths.</code></pre>
<pre class="r"><code># plot the tree
plot(t)</code></pre>
<p><img src="exercise-07a_files/figure-html/tree-1.png" width="672" /></p>
<p>Note the function simulates the <em>complete tree</em>, that is, the
tree containing all lineages that were ever part of the tree (observed
and unobserved).</p>
<p>Run the function a few more times without changing any of the
parameters. You should notice that each tree looks quite different. This
reflects the stochasticity inherent in the birth-death process <span
class="math inline">\(-\)</span> the same combination of birth and death
parameter will result a wide range of possible outcomes.</p>
<blockquote>
<p>Next, increase the birth rate to something like 0.3, and simulate
some new trees. Do you notice any general differences about the trees?
Next, increase the death rate to something like 0.25 and do the same.
What do you notice this time?</p>
</blockquote>
</div>
<div id="the-fossil-sampling-process" class="section level2">
<h2>The fossil sampling process</h2>
<p>Here we’ll simulate fossils, assumming a constant rate of
fossilisation.</p>
<pre class="r"><code># Simulate fossils
rate = 0.3 # poisson sampling rate
f = FossilSim::sim.fossils.poisson(rate = rate, tree = t)
f</code></pre>
<pre><code>##    sp edge       hmin       hmax
## 1   1    1  7.9126891  7.9126891
## 2   3    3  0.4766375  0.4766375
## 3   3    3  2.7507976  2.7507976
## 4   6    6  2.8698887  2.8698887
## 5   7    7  1.7002141  1.7002141
## 6   7    7  0.5138199  0.5138199
## 7   7    7  1.9609156  1.9609156
## 8   9    9  0.6496230  0.6496230
## 9  11   11 20.9641465 20.9641465
## 10 17   17  4.0468324  4.0468324
## ...
## Fossil record with 11 occurrences representing 7 species
## Fossil record not simulated using taxonomy: all speciation events are assumed to be symmetric</code></pre>
<p>Each entry in the fossils data frame corresponds to a fossil sampling
time. <code>sp</code> and <code>edge</code> are the species and edge
labels, respectively. <code>hmin</code> and <code>hmax</code> are the
youngest and oldest sampling ages associated with each fossil,
respectively. Since this function returns exact fossil sampling times
hmin and hmax are equal.</p>
<p><code>FossilSim</code> has a lot of different functions for plotting
the output.</p>
<pre class="r"><code># plot fossil occurrences
plot(f, tree = t)</code></pre>
<p><img src="exercise-07a_files/figure-html/plot-1.png" width="672" /></p>
<pre class="r"><code># plot stratigraphic ranges
plot(f, tree = t, show.ranges = TRUE)</code></pre>
<p><img src="exercise-07a_files/figure-html/plot-2.png" width="672" /></p>
<pre class="r"><code># plot the output
plot(f, tree = t, show.strata = TRUE, strata = 4)</code></pre>
<p><img src="exercise-07a_files/figure-html/plot-3.png" width="672" /></p>
<blockquote>
<p>What happens if you decrease or increase the fossil sampling
rate?</p>
</blockquote>
<p>You can also compare the complete and reconstructed trees.</p>
<pre class="r"><code># plot the reconstructed tree
plot(f, tree = t, reconstructed = TRUE)</code></pre>
<p><img src="exercise-07a_files/figure-html/recon-1.png" width="672" /></p>
</div>
<div id="extant-species-sampling" class="section level2">
<h2>Extant species sampling</h2>
<p>Next we can add include incomplete extant species sampling.</p>
<pre class="r"><code># extant sampling
rho = 0.5

# simulate extant species sampling
f2 = FossilSim::sim.extant.samples(fossils = f, tree = t, rho = rho)

# plot the output
plot(f2, tree = t, extant.col = &quot;red&quot;)</code></pre>
<p><img src="exercise-07a_files/figure-html/extant-1.png" width="672" /></p>
<blockquote>
<p>What happens if you increase or decrease rho?</p>
</blockquote>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
