<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>exercise-06.knit</title>

<script src="site_libs/header-attrs-2.15/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Phylogenetics</a>
</li>
<li>
  <a href="schedule.html">Schedule</a>
</li>
<li>
  <a href="projects.html">Projects</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">




</div>


<p><br />
</p>
<div
id="bayesian-tree-inference-using-the-fossilised-birth-death-process-revbayes"
class="section level1">
<h1>Bayesian tree inference using the fossilised birth-death process
RevBayes</h1>
<p>In this exercise we’ll estimate of a time tree of bears under the
fossilised birth-death (FBD) process. This time the data we’ll use is a
<a href="data/bears_morphology.nex">morphological matrix</a> for 16
living and fossil bears. We’ll also use <a
href="data/bears_taxa.tsv">age information</a> associated with all taxa.
Download these files and check out the contents.</p>
<p>For this exercise create a folder called <code>exercise 6</code>.
Then create two sub-directories: <code>data</code> and
<code>scripts</code>. Recall that for divergence time estimation we need
three model components, so we’ll create the following scripts:</p>
<ol style="list-style-type: decimal">
<li><code>main.Rev</code> for reading the data, setting up the the tree
model and MCMC settings</li>
<li><code>FBD_dating.Rev</code> for the time tree model.</li>
<li><code>Mk.Rev</code> for the substitution model.</li>
<li><code>clock_strict_morpho.Rev</code> for the clock model. Edit the
scripts in a text editor of your choice.</li>
</ol>
<div id="read-in-the-data-and-define-helper-variables"
class="section level2">
<h2>Read in the data and define helper variables</h2>
<p>Start with your <code>main.Rev</code> script. As before, we’ll start
by reading in the data and defining some helper variables.</p>
<pre class="r"><code>taxa &lt;- readTaxonData(&quot;data/bears_taxa.tsv&quot;)
morpho &lt;- readDiscreteCharacterData(&quot;data/bears_morphology.nex&quot;)</code></pre>
<pre class="r"><code>num_taxa &lt;- morpho.ntaxa() # number of taxa
num_branches &lt;- 2 * num_taxa - 2 # number of branches in an rooted tree</code></pre>
<p>Define a set of variables for setting up our MCMC.</p>
<pre class="r"><code>moves    = VectorMoves()
monitors = VectorMonitors()</code></pre>
</div>
<div id="time-tree-model" class="section level2">
<h2>Time tree model</h2>
<p>We will use FBD model as our tree model, which describes the
distribution of most likely trees, given the model parameters
(speciation, extinction, fossil sampling) and fossil ages. Note that
because the fossil ages are data, the FBD model is not strictly a
prior.</p>
<p>Go to your <code>FBD_dating.Rev</code> script to set this up.</p>
<p>Three of the key parameters of the FBD model are speciation rate
(<span class="math inline">\(\lambda\)</span>), extinction rate (<span
class="math inline">\(\mu\)</span>) and fossil sampling rate (<span
class="math inline">\(\psi\)</span>) We will assume these rates are
constant over time and place exponential priors on each of these.</p>
<pre class="r"><code>speciation_rate ~ dnExponential(10)
extinction_rate ~ dnExponential(10)
fossil_samp_rate ~ dnExponential(10)</code></pre>
<p>Each parameter is a stochastic variable drawn from an exponential
distribution with <span class="math inline">\(\delta\)</span> = 10 and a
mean of <span class="math inline">\(1/\delta\)</span> = 0.1.</p>
<p>Next specify the moves of these parameters.</p>
<pre class="r"><code>moves.append( mvScale(speciation_rate, lambda = 0.5, tune = true, weight = 3.0) )
moves.append( mvScale(extinction_rate, lambda = 0.5, tune = true, weight = 3.0) )
moves.append( mvScale(fossil_samp_rate, lambda = 0.5, tune = true, weight = 3.0) )</code></pre>
<p>We might also be interested in diversification and turnover, so let’s
set up a deterministic variable for each of these.</p>
<pre class="r"><code>diversification := speciation_rate - extinction_rate
turnover := extinction_rate/speciation_rate</code></pre>
<p>Next we’ll define the probability of extant species sampling (<span
class="math inline">\(\rho\)</span>). Since we sample all extant bears,
we’ll specify this probability as a constant variable = 1.0.</p>
<pre class="r"><code>rho &lt;- 1.0</code></pre>
<p>Now we’ll specify a prior on the origin time parameter. This
parameter approximates the age of the clade. We’ll use the oldest bear
fossil as a minimum (35.0) and a recent estimate for the age of
carnivores as a maximum (55.0).</p>
<pre class="r"><code># prior and move on the origin
origin_time ~ dnUnif(37.0, 55.0)
moves.append( mvSlide(origin_time, weight=1.0) )</code></pre>
<p>Finally, we’ll specify the FBD model, along with moves on the tree
topology and branch lengths.</p>
<pre class="r"><code>tree ~ dnFBDP(lambda = speciation_rate, mu = extinction_rate, psi = fossil_samp_rate, rho = rho, origin = origin_time, taxa = taxa)

# moves to search tree space
moves.append( mvFNPR(tree, weight = 15.0) )
moves.append( mvCollapseExpandFossilBranch(tree, origin_time, weight = 6.0) )
moves.append( mvNodeTimeSlideUniform(tree, weight = 40.0) )
moves.append( mvRootTimeSlideUniform(tree, origin_time, weight = 5.0) )</code></pre>
<p>We can create clade constraints to monitor the age of specific clades
of interest. Let’s do that for the clade containing living bears.</p>
<pre class="r"><code>clade_extant = clade(&quot;Ailuropoda_melanoleuca&quot;,&quot;Tremarctos_ornatus&quot;,&quot;Melursus_ursinus&quot;,
                    &quot;Ursus_arctos&quot;,&quot;Ursus_maritimus&quot;,&quot;Helarctos_malayanus&quot;,
                    &quot;Ursus_americanus&quot;,&quot;Ursus_thibetanus&quot;)
age_extant := tmrca(tree, clade_extant)</code></pre>
<p>We might also be interested in know how many sampled ancestors are in
our posterior trees, so let’s set up another deterministic variable to
monitor this.</p>
<pre class="r"><code>num_samp_anc := tree.numSampledAncestors()</code></pre>
<div id="accounting-for-fossil-age-uncertainty" class="section level3">
<h3>Accounting for fossil age uncertainty</h3>
<p>Since we practically never know the exact age of any fossils, we need
to account for age uncertainty. The following loop goes through each
fossil in our dataset and assigns a uniform distribution between the
minimum and maximum bounds of the known interval of uncertainty.</p>
<pre class="r"><code>fossils = tree.getFossils()

for(i in 1:fossils.size())
{
    t[i] := tmrca(tree, clade(fossils[i]))

    a_i = fossils[i].getMinAge()
    b_i = fossils[i].getMaxAge()

    F[i] ~ dnUniform(t[i] - b_i, t[i] - a_i)
    F[i].clamp( 0 )
}

# moves on fossil ages
moves.append( mvFossilTimeSlideUniform(tree, origin_time, weight = 5.0) )</code></pre>
</div>
</div>
<div id="strict-clock-model" class="section level2">
<h2>Strict clock model</h2>
<p>For simplicity we’ll use a strict clock model to describe the rate of
character change over time, assuming the rate is constant over time and
across all branches in our tree. <strong>Hint</strong>: you can use your
clock script from the <a href="exercise-04.html#strict">previous
exercise</a>.</p>
<p>Open your <code>clock_strict_morpho.Rev</code> script. We don’t know
the rate of evolution, so as before we’ll use an exponential prior.</p>
<pre class="r"><code>branch_rates ~ dnExponential(10.0)

moves.append( mvScale(branch_rates, lambda = 0.5, tune = true, weight = 3.0) )</code></pre>
<p>This rate will be used for all branches.</p>
</div>
<div id="substitution-model" class="section level2">
<h2>Substitution model</h2>
<p>Next we’ll specify the Mk substitution model in the
<code>Mk.Rev</code> file, which describes the probability of
transitioning from one character state to another. This model is a
generalisation of the Jukes-Cantor model that we’ve encountered
previously.</p>
<p>First we’ll use the <code>fnJC</code> function to define a <strong>Q
matrix</strong> for 2 states (0, 1) since we only have two character
states in our morphological character matrix. Remember that under this
model rates of change between states and state frequencies are
equal.</p>
<pre class="r"><code>Q &lt;- fnJC(2)</code></pre>
<p>Then we’ll define a stochastic node representing a character matrix
and “clamp” that variable to our morphological data. Note that this time
we need to use <code>type = "Standard"</code>.</p>
<pre class="r"><code>seq ~ dnPhyloCTMC(tree = tree, Q = Q, type = &quot;Standard&quot;, branchRates = branch_rates)
seq.clamp(morpho)</code></pre>
</div>
<div id="mcmc-settings" class="section level2">
<h2>MCMC settings</h2>
<p>Back in <code>main.Rev</code> add the following lines to include the
tree, clock and substitution models.</p>
<pre class="r"><code># tree model
source(&quot;scripts/FBD_dating.Rev&quot;)
# clock model
source(&quot;scripts/clock_strict_morpho.Rev&quot;)
# substitution model
source(&quot;scripts/Mk.Rev&quot;)</code></pre>
<p>The MCMC set up is very similar to before. First we create the
variable <code>mymodel</code>.</p>
<pre class="r"><code>mymodel = model(tree)</code></pre>
<p>Then we define a set of monitors to capture the output.</p>
<pre class="r"><code># parameters printed to file
monitors.append( mnModel(filename = &quot;output/bears_FBD.log&quot;, printgen = 10, exclude = [&quot;F&quot;]) )
# trees printed to file
monitors.append( mnFile(filename = &quot;output/bears_FBD.trees&quot;, printgen = 10, tree) )
# parameter values printed to screen during the MCMC
monitors.append( mnScreen(printgen = 10, age_extant, num_samp_anc, origin_time) )</code></pre>
<p><code>exclude = ["F"]</code> here means fossil ages will be excluded
from the log file - I’ve done this because the way RevBayes output this
information is a bit confusing, so we’ll ignore that for the moment.</p>
<p>Finally, we’ll set up the MCMC run using the <code>mcmc</code>
function, specifying our model, the vector of monitors and the vector of
moves and run the chain for 10000 generations.</p>
<pre class="r"><code>mymcmc = mcmc(mymodel, monitors, moves)
mymcmc.run(generations = 10000, tuningInterval = 1000)</code></pre>
<p>Note that this time we’ve adding the argument
<code>tuningInterval = 1000</code>. This tells the chain how long to
spend optimising the MCMC moves. Different datasets behave differently
<span class="math inline">\(-\)</span> the above option partly automates
the selection of move parameters that maximise efficiency of the
MCMC.</p>
<p>This might take a while run pretty quickly. Meanwhile, you could move
on to the <a href="#next">Next tasks</a>.</p>
</div>
<div id="evaluating-the-output" class="section level2">
<h2>Evaluating the output</h2>
<p>As before, open your <code>.log</code> file in Tracer.<br />
It should look something like this.</p>
<p><img src="trace5.png" width="85%" style="padding:10px" /></p>
<blockquote>
<p>Explore the output. Can you identify the different parameters? Has
the analysis converged?</p>
</blockquote>
<blockquote>
<p>What is the age of the most common recent ancestor of bears?</p>
</blockquote>
<p>Next let’s generate a summary tree.</p>
<pre class="r"><code>trace = readTreeTrace(&quot;output/bears_FBD.trees&quot;)
mccTree(trace, file = &quot;output/bears_FBD.mcc.tre&quot;, positiveBranchLengths = TRUE)</code></pre>
<p>Once you have your summary tree open it in FigTree and play around
with the settings. See if you can get something like this.</p>
<p><img src="mcc_bears_fbd.png" width="85%" style="padding:10px" /></p>
<blockquote>
<p>Which candiates on your tree do you think might be sampled
ancestors?</p>
</blockquote>
</div>
<div id="next" class="section level2">
<h2>Next tasks</h2>
<p>Some extra things to try. <strong>Make sure you rename the output
files so you don’t overwrite the files you generated above!</strong></p>
<div id="add-molecular-sequence-data" class="section level3">
<h3>Add molecular sequence data</h3>
<p>Try adding the molecular sequence alignment from the previous
exercise.</p>
<p>There’s just a few extra steps we need to add when we read in the
data, at the beginning of <code>main.Rev</code>.</p>
<p>First add the line that reads in the sequence data and create a
constant variable <code>cytb</code>.</p>
<pre class="r"><code>taxa &lt;- readTaxonData(&quot;data/bears_taxa.tsv&quot;)
morpho &lt;- readDiscreteCharacterData(&quot;data/bears_morphology.nex&quot;)
cytb &lt;- readDiscreteCharacterData(&quot;data/bears_cytb.nex&quot;)</code></pre>
<p>Next we need to make sure the datasets are cross compatible because
we don’t have molecular sequence data for all species.</p>
<pre class="r"><code>cytb.addMissingTaxa( taxa )
morpho.addMissingTaxa( taxa )</code></pre>
<p>That’s it! Now you can go ahead and set up the
[GTR]{exercise-05.html#GTR} model as before.</p>
<blockquote>
<p>What differences do you notice in your output?</p>
</blockquote>
</div>
<div id="running-the-analysis-under-the-fbd" class="section level3">
<h3>Running the analysis under the FBD</h3>
<p>Can you try running the analysis <a
href="exercise-05.html#prior">under the prior</a>? Note that its not
strictly “under the prior” because the fossil ages are data but RevBayes
doesn’t consider the FBD as part of the likelihood, so you can still
choose the regular option for running the analysis under the prior to
distinguish the signal coming from the FBD/fossil age data vs. the FBD
combined with the sequence/morpho data.</p>
<blockquote>
<p>Compare you results to the posterior obtained using in the previous
steps. What information do we gain from the morphological / molecular
data?</p>
</blockquote>
<p>A complete set of scripts for this exercise can be downloaded <a
href="https://fau-paleo.github.io/apw_2022/data/7_phylogenetics/exercise-RB03-scripts.zip">here</a>.</p>
</div>
<div id="acknowledgements" class="section level3">
<h3>Acknowledgements</h3>
<p>This exercise is adapted from the one described <a
href="https://revbayes.github.io/tutorials/fbd/fbd_specimen.html">here</a>
by Tracy Heath, April Wright, Walker Pett and <a
href="https://hal.inria.fr/PGE/hal-02536394">here</a> by Barido-Sottani
et al. </p>
</div>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
