<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>exercise-01.knit</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Phylogenetics</a>
</li>
<li>
  <a href="schedule.html">Schedule</a>
</li>
<li>
  <a href="projects.html">Projects</a>
</li>
<li>
  <a href="exercise-06a.html">Next exercise</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">




</div>


<p><br />
</p>
<div id="tree-building-using-parsimony" class="section level1">
<h1>Tree building using parsimony</h1>
<p>The purpose of this short exercise is to show you how quick and
simple it can be to generate a tree using R.</p>
<p>We’ll be using a morphological matrix with representatives of
Ediacaran and Cambrian animals generated by the fab palaeobiologist <a
href="https://oumnh.ox.ac.uk/people/dr-frankie-dunn">Francis Dunn</a>.
The original data set was downloaded from <a
href="https://research-information.bris.ac.uk/en/studentTheses/growth-and-development-in-the-ediacaran-macrobiota">here</a>
and you can read the published results in <a
href="https://www.science.org/doi/epdf/10.1126/sciadv.abe2741">Science
Advances</a>. The purpose of this tree was to try and establish this
phylogenetic affinity of the enigmatic <a
href="https://en.wikipedia.org/wiki/Charnia"><em>Charnia
masoni</em></a>, which is only known from the Ediacaran.</p>
<p>First install the necessary R packages.</p>
<pre class="r"><code>install.packages(&quot;ape&quot;)
install.packages(&quot;phangorn&quot;)</code></pre>
<p><code>ape</code> is a very widely used phyologenetics package that
contains (among many others) functions for reading, writing and plotting
phylogenetic trees and data. <code>phangorn</code> is a package that
contains functions for straightforward tree building.</p>
<p>Download the character data from <a
href="data/Frankie-data.nex">here</a>. Nexus is a common format used to
store phylogenetic character data. Next we’ll read the data and convert
it into the format used by phangorn.</p>
<pre class="r"><code># read in the character matrix in ape 
dat = ape::read.nexus.data(file = &quot;data/Frankie-data.nex&quot;)

# identify character states
chars = unique(unlist(unique(dat)))

# &amp; convert to phanghorn object
dat = phangorn::phyDat(dat, type = &quot;USER&quot;, levels = chars)</code></pre>
<p>Next we need to provide <code>phangorn</code> with a starting tree.
We can do this using an approach called
<strong>neighbour-joining</strong>. This simply calculates the
evolutionary distance between the taxa in your matrix (using the
<code>dist.hamming</code> function) and then groups taxa together based
on their relative distances (using the <code>NJ</code> function).</p>
<pre class="r"><code># create a neighbour joining starting tree
dm = phangorn::dist.hamming(dat)
tree = phangorn::NJ(dm)</code></pre>
<p>Then we’ll use the <code>phangorn</code> function
<code>optim.parsimony</code> to find the most parsimonious tree.</p>
<pre class="r"><code># estimate the parsimony tree
mp = phangorn::optim.parsimony(tree, dat)</code></pre>
<pre><code>## Final p-score 305 after  1 nni operations</code></pre>
<blockquote>
<p>How would you interpret the number output by this function?</p>
</blockquote>
<p>Finally, we’ll can specify an outgroup to root the tree. We’ll use
the fungus <em>Laccaria</em> because we know this is more distantly
related to all other taxa in our tree than they are to eachother.</p>
<pre class="r"><code># identify the outgroup
outgroup = &quot;Laccaria&quot;

# root the tree
mp = ape::root(mp, outgroup = outgroup, resolve.root = TRUE)

plot(mp)</code></pre>
<p><img src="exercise-01_files/figure-html/plot-1.png" width="672" /></p>
<blockquote>
<p>Based on this tree how would you interpret the phylogenetic position
of <em>Charnia</em>? This is a bit tricky if you don’t know all the
Latin names, but could you say whether it was a stem or crown group
animal?</p>
</blockquote>
<div id="other-resources-for-buidling-trees-using-parsimony"
class="section level2">
<h2>Other resources for buidling trees using parsimony</h2>
<p>Because there are a large number of possible trees even for a small
number of taxa, finding the most parsimonous tree is actually really
algorithmically challenging. Many people use the open source software <a
href="http://www.lillo.org.ar/phylogeny/tnt/">TNT</a>, which can speed
up the tree search. You can also tweek your parsimony analysis based on
assumptions you want to make about your character data. For example,
characters can be ordered (e.g. you can allow character state changes
from 0 to 1 or from 1 to 2 but prevent changes from 0 to 2). You can
also assign different weights to different characters if you believe
some morphological characters are more informative about taxonomy.</p>
<p>We won’t explore parsimony any further in this course but it remains
widely used for the anaylsis of morphologcial character data and is
sometimes much faster than alternative tree-building approaches. It can
be a good way to obtain an initial look at your data/tree before or
while running more computationally intensive analyses.</p>
<p>A complete script for this exercise can be downloaded <a
href="scripts/parsimony_example.R">here</a>.</p>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
