<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>FossilSimShiny.knit</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Phylogenetics</a>
</li>
<li>
  <a href="schedule_part_1.html">Part 1</a>
</li>
<li>
  <a href="schedule_part_2.html">Part 2</a>
</li>
<li>
  <a href="projects.html">Projects</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">




</div>


<p><br />
</p>
<div id="exploring-model-assumptions-using-the-fossilsimshiny-app"
class="section level1">
<h1>Exploring model assumptions using the FossilSimShiny app</h1>
<p>This exercise will walk you through key assumptions of the fossilized
birth-death (FBD) process (and related models) using the FossilSimShiny
app, which uses the R packages TreeSim and FossilSim for simulation.</p>
<p>The FBD process is a joint model of diversification and sampling,
first described in Stadler (2010). It can be used to calculate the
probability of observing a reconstructed phylogeny with fossils as
extinct leaves or sampled ancestors. The model is statistically coherent
for the analysis of phylogenetic and fossil data because it incorporates
explicit assumptions about the underlying data generating processes.
This means we can forward simulate, as well as perform inference, under
the model, which is a major advantage for understanding the model
properties and assumptions. We can also use the app to explore other
models of fossil sampling, as well as models of taxonomy and
speciation.</p>
<p>Install the necessary R packages.</p>
<pre class="r"><code>install.packages(&quot;TreeSim&quot;)
install.packages(&quot;FossilSim&quot;)
install.packages(&quot;FossilSimShiny&quot;)</code></pre>
<p>To launch the FossilSimShiny app on your computer simply run the
following in R.</p>
<pre class="r"><code>FossilSimShiny::launchFossilSimShiny()</code></pre>
<p>This should open the app in your default browser and look like the
image below.</p>
<p><img src="shiny_01.png" width="100%" /></p>
<p>On the upper left there are three menus used for different aspects of
the simulation: <strong>Tree</strong>, <strong>Taxonomy</strong> and
<strong>Fossils</strong>. On the lower left is the
<strong>Appearance</strong> menu where you can change how the simulated
data is displayed. On the upper right you can also use the drop down
menu to toggle between display options.</p>
<div id="the-diversification-process" class="section level2">
<h2>The diversification process</h2>
<p>First we’ll simulate trees, and explore the role of the birth and
death parameters. Navigate to the <strong>Tree</strong> menu.</p>
<p>We’ll simulate birth-death trees conditioned on the number of tips
<span class="math inline">\(n\)</span> (<strong>Tip number</strong>).
The underlying function simulates birth and death events forward in time
until there are <span class="math inline">\(n\)</span> co-existing
lineages.</p>
<p>Let’s give it a go.</p>
<p><img src="shiny_02.png" width="100%" /></p>
<p>The function simulates the <em>complete tree</em>, that is, the tree
containing all lineages that were ever part of the tree. This includes
all extinct lineages and the <span class="math inline">\(n\)</span>
extant samples at the end of the process (<span class="math inline">\(t
= 0\)</span>). The black diamonds indicated the extant samples here.
Learn more about the complete and reconstructed trees <a
href="FossilSimShiny.html#reconstructed">below</a>.</p>
<p>Run the function a few more times without changing any of the
parameters. You should notice that each tree looks quite different. This
reflects the stochasticity inherent in the birth-death process <span
class="math inline">\(-\)</span> the same combination of speciation and
extinction parameters will result a wide range of possible outcomes.
These rate parameters are often referred to as the birth (<span
class="math inline">\(\lambda\)</span>) and death (<span
class="math inline">\(\mu\)</span>) rates.</p>
<blockquote>
<p>Next, increase the <strong>Speciation rate</strong> to something like
0.3, and simulate some new trees. Do you notice any general differences?
Next, increase the <strong>Extinction rate</strong> to something like
0.25 and do the same. What do you notice this time?</p>
</blockquote>
<p>Before we move on, see if you can built up some intuition for the
interaction between the speciation and extinction rates.</p>
<p><em>To simulate trees using the same approach in R you can use
the</em> <code>TreeSim</code> <em>package function</em>
<code>sim.taxa.bd</code>.</p>
</div>
<div id="the-fossil-sampling-process" class="section level2">
<h2>The fossil sampling process</h2>
<p>Next we’ll simulate fossils. Navigate to the <strong>Fossils</strong>
menu.</p>
<div id="uniform-sampling" class="section level3">
<h3>Uniform sampling</h3>
<p>We’ll start with constant fossil sampling. This is the assumption
made by the original FBD model described by Stadler (2010).</p>
<p>Start with the default parameters and select <strong>Simulate
fossils</strong>.</p>
<p><img src="shiny_03.png" width="100%" /></p>
<p>Each diamond plotted along the tree branches represents a fossil
occurrence. Here, an “occurrence” represents a singular observation of a
lineage at a single point in time. In reality, a lineage can be sampled
multiple times at a given time (e.g., if multiple specimens are sampled
at the same locality, sometimes known as abundance data) but the FBD
model can’t currently account for this type of information. This
reflects the fact that the model assumes a <a
href="https://en.wikipedia.org/wiki/Poisson_distribution">Poisson
sampling process</a>.</p>
<p>At the top of the screen you can see details about your simulation,
including the number of simulated fossil samples. You should notice that
if you simulate fossils using the same sampling rate, each simulation
will generate a different number of samples. The sampling parameter
(<span class="math inline">\(\psi\)</span>) is the rate at which fossils
are sampled along branches in the phylogeny, and encompasses all
processes that lead to fossilisation and sampling.</p>
<blockquote>
<p>What happens if you decrease or increase the <strong>Sampling
rate</strong>?</p>
</blockquote>
<p>Before we move on, toggle between the <strong>Tree</strong> and
<strong>Fossils</strong> menus to see if you can built up some intuition
for the interaction between the speciation, extinction <em>and</em>
sampling rates.</p>
<p>The app has a lot of options for changing the output that is
displayed. Scroll down to the <strong>Appearance</strong> menu to view
the options.</p>
<blockquote>
<p>Deselect <strong>Show tree</strong>. What do you notice about the
distribution of fossils through time when you change the speciation,
extinction and sampling rates?</p>
</blockquote>
<p>Through this approach you should begin to get a sense of the
interaction between the speciation, extinction, sampling rates and the
distribution of fossils. This interaction is why the distribution of
fossil sampling times is informative about the model parameters.</p>
<p><em>To simulate fossils using the same approach in R you can use
the</em> <code>FossilSim</code> <em>package function</em>
<code>sim.fossils.poisson</code>.</p>
</div>
<div id="reconstructed" class="section level3">
<h3>The complete vs reconstructed tree</h3>
<p>Stay in the <strong>Appearance menu</strong> for now and select
<strong>Show reconstructed tree</strong>. This will show you the
contrast between the complete and reconstructed trees.</p>
<p><img src="shiny_04.png" width="100%" /></p>
<p>After adding fossil sampling, the complete tree will be the outcome
of the diversification and sampling processes and will include all
observed and unobserved lineages, plus all fossil and extant samples. In
contrast, the reconstructed tree is the one we can estimate from our
observed samples and is the tree object that we use to calculate the
probability under the FBD model during inference.</p>
<blockquote>
<p>What differences do you notice between your complete and
reconstructed trees?</p>
</blockquote>
</div>
<div id="stratigraphic-ranges" class="section level3">
<h3>Stratigraphic ranges</h3>
<p>In the appearance menu you can also choose to display stratigraphic
ranges. If you select <strong>Show ranges</strong> and deselect
<strong>Show all occurrences</strong> this will show the complete or
reconstructed tree with stratigraphic ranges - that is, the interval
between the first and last appearance of a lineage in the fossil
record.</p>
<p><img src="shiny_05.png" width="100%" /></p>
<p>The purple lines between the first and last appearance represent the
stratigraphic ranges. The first and last appearances are represented by
the black diamonds as before. If a lineage is sampled only once (i.e.,
the first and last appearance are the same), the lineage is represented
by a single occurrence.</p>
<blockquote>
<p>What do you notice about the range durations if you change the
speciation, extinction or rate parameters?</p>
</blockquote>
<p>The FBD range model, which can be used to calculate the probability
of a tree with ranges, was described in Stadler et al. (2018).</p>
<p>Note that by default the simulation functions assume each branch
represents a unique taxon, which is equivalent to assuming all
speciation is fully symmetric or bifurcating. See section <a
href="FossilSimShiny.html#taxonomy">Taxonomy</a> to learn more about
this.</p>
</div>
<div id="time-dependent-sampling" class="section level3">
<h3>Time-dependent sampling</h3>
<p>Return to the <strong>Fossils</strong> menu and select the
<strong>Time-dependent</strong> option.</p>
<p>Next we’ll simulate fossils with rate variation through time. Using
this option, discrete intervals are assigned a separate rate. This is
equivalent to the assumptions made by the FBD skyline model
(Gavryushkina, 2014).</p>
<p>Go to the <strong>Interval type</strong> drop down menu and select
<strong>even</strong>. Use the <strong>Number of intervals</strong>
slider to select the number of intervals you’d like. Rates are simulated
from a lognormal distribution with user specified parameters
<strong>Mean sampling</strong> rate and <strong>Variance</strong>.</p>
<p>Scroll down to the <strong>Appearance</strong> menu and select
<strong>Show strata used in simulating fossils</strong></p>
<p><img src="shiny_06.png" width="100%" /></p>
<p>You should be able to see variation in the density of fossils through
time, although this will depend on the magnitude of the difference in
rates between intervals, as well as the total number of lineages that
occur in each interval.</p>
<blockquote>
<p>Experiment with the other settings associated with time-dependent
sampling and deselect <strong>Show tree</strong> again from the
<strong>Appearance menu</strong>. What do you notice about the
distribution of fossils through time with different rates of specation,
extinction or variation in sampling?</p>
</blockquote>
<p><em>To simulate fossils using the same approach in R you can use
the</em> <code>FossilSim</code> <em>package function</em>
<code>sim.fossils.intervals</code>.</p>
</div>
</div>
<div id="taxonomy" class="section level2">
<h2>Taxonomy</h2>
<p>Navigate to the <strong>Taxonomy</strong> menu. Here we can
incorporate explicit assumptions about taxonomy or speciation mode.
There are three main modes of speciation that can be considered in the
context of birth-death process models: * asymmetric or budding
speciation, where a lineage splitting event gives rise to a single new
species * symmetric or bifurcating specation, where a a lineage
splitting event gives rise to two new species and results in the
extinction of the ancestral species * anagenesis, where speciation
occurs along internal branches and there are no associated lineage
splitting events</p>
<p>The assumptions made by birth-death processes about speciation mode
are not immediately obvious. In standard birth-death models,
mathematically we assume that speciation events are asymmetric, since
lineage splitting events give rise to a single new lineage. This isn’t
always clear, since in phylogenetics we tend to plot trees as full
bifurcating, as if each lineage splitting events gives rise to two new
lineages. Meanwhile, in paleontology taxa defined on the basis of
morphology, and may be the result of symmetric, asymmetric or anagenetic
speciation. For more about the implications of this see Silvestro et
al. (2014).</p>
<p>Under the original FBD model we don’t make any assumptions about how
fossil samples are linked to species or taxa through time, but it
becomes important to consider when we want to infer trees using
stratigraphic ranges. Stadler et al. (2018) introduced the possibility
that speciation could via any of the above three modes.</p>
<p>By default the FossilSimShiny app assumes that all speciation events
are symmetric. Thus, when we plot the ranges, each range is associated
with a single branch in a bifurcating tree.</p>
<p>If you navigate to the <strong>Taxonomy menu</strong> we can
introduce alternative speciation modes. The <strong>Probability of
symmetric speciation</strong> is the probability that a given lineage
splitting event will be symmetric or bifurcating (<span
class="math inline">\(\beta\)</span>) (otherwise it is asymmetric or
budding with probability <span class="math inline">\(\beta-1\)</span>).
The <strong>Rate of anagenesis</strong> is simply the rate of anagensis
along branches (<span class="math inline">\(\lambda_a\)</span>).
Simulate taxonomy starting with the default values.</p>
<p><img src="shiny_07.png" width="100%" /></p>
<p>The <strong>taxonomy</strong> plot shows how speciation events are
distributed across the tree. You can use the legend to identify the
symbol associated with each speciation mode.</p>
<blockquote>
<p>Experiment with the <strong>Probability of symmetric
speciation</strong> and <strong>Rate of anagenesis</strong> parameters.
Do the numbers of each speciation make sense based on the input
parameters you use?</p>
</blockquote>
<p>If you select <strong>tree+fossils</strong> from the top right, you
can also see the fossils coloured according to their taxonomy.</p>
<p><img src="shiny_08.png" width="100%" /></p>
<blockquote>
<p>What challenges do you percieve working with range data and
conventional phylogenetic approaches that do not take speciation mode
into account? Or when comparing parameter estimates based on fossils
vs. phylogenies?</p>
</blockquote>
<blockquote>
<p>If you have time you can return and experiment with the
<strong>Depth-dependent</strong> and <strong>Lineage-dependent</strong>
sampling.</p>
</blockquote>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
